File: API/2759_Coin_Flipping/CS/CoinFlippingStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Randomized coin flipping strategy that alternates between buying and selling based on a pseudo-random generator.
/// Mimics the original MetaTrader expert advisor by opening a single position at a time with symmetric risk controls.
/// </summary>
public class CoinFlippingStrategy : Strategy
{
	private readonly StrategyParam<decimal> _riskPercent;
	private readonly StrategyParam<int> _takeProfitPips;
	private readonly StrategyParam<int> _stopLossPips;
	private readonly StrategyParam<DataType> _candleType;

	private Random _random;
	private decimal _priceStep;
	private decimal _takeProfitDistance;
	private decimal _stopLossDistance;

	/// <summary>
	/// Portfolio share allocated to every trade in percent.
	/// </summary>
	public decimal RiskPercent
	{
		get => _riskPercent.Value;
		set => _riskPercent.Value = value;
	}

	/// <summary>
	/// Take profit distance measured in pips.
	/// </summary>
	public int TakeProfitPips
	{
		get => _takeProfitPips.Value;
		set => _takeProfitPips.Value = value;
	}

	/// <summary>
	/// Stop loss distance measured in pips.
	/// </summary>
	public int StopLossPips
	{
		get => _stopLossPips.Value;
		set => _stopLossPips.Value = value;
	}

	/// <summary>
	/// Candle type used for scheduling trade attempts.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Initialize strategy parameters.
	/// </summary>
	public CoinFlippingStrategy()
	{
		_riskPercent = Param(nameof(RiskPercent), 2m)
			.SetGreaterThanZero()
			.SetDisplay("Risk %", "Portfolio percentage allocated per trade", "Risk Management")
			.SetCanOptimize(true)
			.SetOptimize(1m, 10m, 1m);

		_takeProfitPips = Param(nameof(TakeProfitPips), 20)
			.SetGreaterThanZero()
			.SetDisplay("Take Profit (pips)", "Target distance expressed in pips", "Risk Management")
			.SetCanOptimize(true)
			.SetOptimize(10, 50, 5);

		_stopLossPips = Param(nameof(StopLossPips), 10)
			.SetGreaterThanZero()
			.SetDisplay("Stop Loss (pips)", "Protective stop distance expressed in pips", "Risk Management")
			.SetCanOptimize(true)
			.SetOptimize(5, 30, 5);

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(1).TimeFrame())
			.SetDisplay("Candle Type", "Candle type used for trade timing", "Data");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		// Reset cached state when the strategy is reset.
		_random = null;
		_priceStep = 0m;
		_takeProfitDistance = 0m;
		_stopLossDistance = 0m;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		// Seed the pseudo-random generator similarly to the MQL expert.
		_random = new Random(Environment.TickCount);

		// Determine price step information for translating pips into price units.
		_priceStep = Security?.PriceStep ?? 1m;
		if (_priceStep <= 0m)
			_priceStep = 1m;

		_takeProfitDistance = TakeProfitPips * _priceStep;
		_stopLossDistance = StopLossPips * _priceStep;

		var takeProfitUnit = _takeProfitDistance > 0m
			? new Unit(_takeProfitDistance, UnitTypes.Price)
			: new Unit();

		var stopLossUnit = _stopLossDistance > 0m
			? new Unit(_stopLossDistance, UnitTypes.Price)
			: new Unit();

		// Attach protective orders so every position has an exit target and stop.
		StartProtection(takeProfitUnit, stopLossUnit);

		// Subscribe to candle data to trigger decision making once per bar.
		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(ProcessCandle).Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		// Only use completed candles to avoid duplicate executions while a bar is forming.
		if (candle.State != CandleStates.Finished)
			return;

		if (!IsFormedAndOnlineAndAllowTrading())
			return;

		// The strategy maintains at most one position at a time.
		if (Position != 0)
			return;

		if (_random == null)
			return;

		var entryPrice = candle.ClosePrice;
		if (entryPrice <= 0m)
			return;

		var volume = CalculateOrderVolume(entryPrice);
		if (volume <= 0m)
			return;

		var isBuy = _random.Next(0, 2) == 0;
		if (isBuy)
		{
			LogInfo($"Opening long position at {entryPrice} with volume {volume}.");
			BuyMarket(volume);
		}
		else
		{
			LogInfo($"Opening short position at {entryPrice} with volume {volume}.");
			SellMarket(volume);
		}
	}

	private decimal CalculateOrderVolume(decimal entryPrice)
	{
		var balance = Portfolio?.CurrentValue ?? 0m;
		if (balance <= 0m)
			return 0m;

		var riskAmount = balance * RiskPercent / 100m;
		if (riskAmount <= 0m)
			return 0m;

		var stopDistance = _stopLossDistance;
		if (stopDistance <= 0m)
		{
			stopDistance = StopLossPips * _priceStep;
		}

		if (stopDistance <= 0m)
			return 0m;

		// Risk per unit equals the stop distance; divide to get the number of contracts.
		var rawVolume = riskAmount / stopDistance;
		var volume = NormalizeVolume(rawVolume);

		if (volume <= 0m)
		{
			volume = Volume > 0m ? Volume : 1m;
			volume = NormalizeVolume(volume);
		}

		return volume;
	}

	private decimal NormalizeVolume(decimal volume)
	{
		if (volume <= 0m)
			return 0m;

		if (Security?.VolumeStep is decimal step && step > 0m)
		{
			volume = Math.Floor(volume / step) * step;
		}

		if (Security?.MinVolume is decimal minVolume && minVolume > 0m && volume < minVolume)
		{
			volume = minVolume;
		}

		if (Security?.MaxVolume is decimal maxVolume && maxVolume > 0m && volume > maxVolume)
		{
			volume = maxVolume;
		}

		return volume;
	}
}

File: API/2701-2800/2759_Coin_Flipping/PY/coin_flipping_strategy.py
import clr
import random

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan

from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Strategies import Strategy


class coin_flipping_strategy(Strategy):
    def __init__(self):
        super(coin_flipping_strategy, self).__init__()

        self._risk_percent = self.Param("RiskPercent", 2.0)
        self._take_profit_pips = self.Param("TakeProfitPips", 5000)
        self._stop_loss_pips = self.Param("StopLossPips", 3000)
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromDays(1)))

        self._rng = None
        self._price_step = 1.0
        self._tp_dist = 0.0
        self._sl_dist = 0.0
        self._entry_price = 0.0
        self._stop_price = None
        self._take_price = None

    @property
    def CandleType(self):
        return self._candle_type.Value

    @CandleType.setter
    def CandleType(self, value):
        self._candle_type.Value = value

    @property
    def RiskPercent(self):
        return self._risk_percent.Value

    @property
    def TakeProfitPips(self):
        return self._take_profit_pips.Value

    @property
    def StopLossPips(self):
        return self._stop_loss_pips.Value

    def OnStarted2(self, time):
        super(coin_flipping_strategy, self).OnStarted2(time)

        self._rng = random.Random()

        sec = self.Security
        self._price_step = float(sec.PriceStep) if sec is not None and sec.PriceStep is not None else 1.0
        if self._price_step <= 0:
            self._price_step = 1.0

        self._tp_dist = self.TakeProfitPips * self._price_step
        self._sl_dist = self.StopLossPips * self._price_step

        self._entry_price = 0.0
        self._stop_price = None
        self._take_price = None

        subscription = self.SubscribeCandles(self.CandleType)
        subscription.Bind(self._process_candle).Start()

        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawOwnTrades(area)

    def _process_candle(self, candle):
        if candle.State != CandleStates.Finished:
            return

        high = float(candle.HighPrice)
        low = float(candle.LowPrice)
        close = float(candle.ClosePrice)

        if self.Position > 0:
            if self._stop_price is not None and low <= self._stop_price:
                self.SellMarket()
                self._reset_targets()
            elif self._take_price is not None and high >= self._take_price:
                self.SellMarket()
                self._reset_targets()
        elif self.Position < 0:
            if self._stop_price is not None and high >= self._stop_price:
                self.BuyMarket()
                self._reset_targets()
            elif self._take_price is not None and low <= self._take_price:
                self.BuyMarket()
                self._reset_targets()

        if self.Position != 0:
            return

        if self._rng is None:
            return

        if close <= 0:
            return

        is_buy = self._rng.randint(0, 1) == 0

        if is_buy:
            self.BuyMarket()
            self._entry_price = close
            self._stop_price = close - self._sl_dist if self._sl_dist > 0 else None
            self._take_price = close + self._tp_dist if self._tp_dist > 0 else None
        else:
            self.SellMarket()
            self._entry_price = close
            self._stop_price = close + self._sl_dist if self._sl_dist > 0 else None
            self._take_price = close - self._tp_dist if self._tp_dist > 0 else None

    def _reset_targets(self):
        self._entry_price = 0.0
        self._stop_price = None
        self._take_price = None

    def OnReseted(self):
        super(coin_flipping_strategy, self).OnReseted()
        self._rng = None
        self._price_step = 1.0
        self._tp_dist = 0.0
        self._sl_dist = 0.0
        self._entry_price = 0.0
        self._stop_price = None
        self._take_price = None

    def CreateClone(self):
        return coin_flipping_strategy()
